library/src/main/java/kwasm/format/text/Tokenizer.kt

```diff
@@ -0,0 +1,135 @@
+import kwasm.format.text.token.isKeyword
+import kwasm.format.text.token.isOpenParen
+import kwasm.format.text.token.isReserved
+import kwasm.format.text.token.isStringLiteral
+import kwasm.format.text.token.toFloatLiteral
+import kwasm.format.text.token.toIdentifier
+import kwasm.format.text.token.toIntegerLiteral
+import kwasm.format.text.token.toKeyword
+import kwasm.format.text.token.toParen
+import kwasm.format.text.token.toReserved
+import kwasm.format.text.token.toStringLiteral
+import kwasm.format.text.whitespace.Comment
+import kwasm.format.text.whitespace.Format
+import java.io.Reader
+import java.io.StringReader
+
+/**
+ * A tokenizer capable of splitting a raw text-format WASM file into its component tokens.
+ *
+ * From [the docs](https://webassembly.github.io/spec/core/text/lexical.html#tokens):
+ *
+ * The character stream in the source text is divided, from left to right, into a sequence of
+ * tokens, as defined by the following grammar.
+ *
+ * ```
+ *   token      ::= keyword | uN | sN | fN | string | id | ‘(’ | ‘)’ | reserved
+ *   keyword    ::= (‘a’ | … |‘z’) idchar* (if occurring as a literal terminal in the grammar)
+ *   reserved   ::= idchar+
+ * ```
+ *
+ * Tokens are formed from the input character stream according to the *longest match rule*. That is,
+ * the next token always consists of the longest possible sequence of characters that is recognized
+ * by the above lexical grammar. Tokens can be separated by white space, but except for strings,
+ * they cannot themselves contain whitespace.
+ */
+class Tokenizer {
+    /** Tokenizes source code from the provided [source] [Reader]. */
+    fun tokenize(
+        source: Reader,
+        context: ParseContext? = ParseContext("Unknown File", 1, 1)
+    ): List<Token> {
+        val rawTokens = source.use { Comment.stripComments(it.readText(), context) }.tokens
+        return rawTokens.flatMap(::buildActualTokensFrom)
+    }
+
+    /** Tokenizes source code from the provided [source] [String]. */
+    fun tokenize(
+        source: String,
+        context: ParseContext? = ParseContext("Unknown File", 1, 1)
+    ): List<Token> = tokenize(StringReader(source), context)
+
+    // TODO: it might be worth considering making a lot of these functions `inline fun`s.
+    private fun buildActualTokensFrom(token: RawToken): List<Token> = when {
+        token.sequence.isEmpty() -> emptyList()
+        Format.PATTERN.get().matchEntire(token.sequence) != null -> emptyList()
+        token.isKeyword() -> listOf(token.toKeyword())
+        token.isIntegerLiteral() -> listOf(token.toIntegerLiteral())
+        token.isFloatLiteral() -> listOf(token.toFloatLiteral())
+        token.isStringLiteral() -> listOf(token.toStringLiteral())
+        token.isIdentifier() -> listOf(token.toIdentifier())
+        token.isOpenParen() || token.isClosedParen() -> listOf(token.toParen())
+        token.isReserved() -> listOf(token.toReserved())
+        else -> {
```

Comment:
So when parsing the raw token, if it could parse the token to exactly one type, it will be set? Otherwise this else figures out the longest match?

Reply:
Yep!